src/coreclr/src/gc/unix/cgroup.cpp

```diff
@@ -59,96 +67,116 @@ class CGroup
-        result = ReadMemoryValueFromFile(mem_limit_filename, val);
-        free(mem_limit_filename);
-        return result;
+        if (s_cgroup_version == 0)
+            return false;
+        else if (s_cgroup_version == 1)
+            return GetCGroupMemoryLimit(val, CGROUP1_MEMORY_LIMIT_FILENAME);
+        else if (s_cgroup_version == 2)
+            return GetCGroupMemoryLimit(val, CGROUP2_MEMORY_LIMIT_FILENAME);
+        else
+        {
+            assert(!"Unknown cgroup version.");
+            return false;
+        }
     }
 
     static bool GetPhysicalMemoryUsage(size_t *val)
     {
-        char *mem_usage_filename = nullptr;
-        bool result = false;
-        uint64_t temp = 0;
+        if (s_cgroup_version == 0)
+            return false;
+        else if (s_cgroup_version == 1)
+            return GetCGroupMemoryUsage(val, CGROUP1_MEMORY_USAGE_FILENAME);
+        else if (s_cgroup_version == 2)
+            return GetCGroupMemoryUsage(val, CGROUP2_MEMORY_USAGE_FILENAME);
+        else
+        {
+            assert(!"Unknown cgroup version.");
+            return false;
+        }
+    }
 
-        if (s_memory_cgroup_path == nullptr)
-            return result;
+    static bool GetCpuLimit(uint32_t *val)
+    {
+        if (s_cgroup_version == 0)
+            return false;
+        else if (s_cgroup_version == 1)
+            return GetCGroup1CpuLimit(val);
+        else if (s_cgroup_version == 2)
+            return GetCGroup2CpuLimit(val);
+        else
+        {
+            assert(!"Unknown cgroup version.");
+            return false;
+        }
+    }
 
-        size_t len = strlen(s_memory_cgroup_path);
-        len += strlen(MEM_USAGE_FILENAME);
-        mem_usage_filename = (char*)malloc(len+1);
-        if (mem_usage_filename == nullptr)
-            return result;
+private:
+    static int FindCGroupVersion()
+    {
+        int version = 0;
+        char *line = nullptr;
+        size_t lineLen = 0;
+        char *filesystemType = nullptr;
 
-        strcpy(mem_usage_filename, s_memory_cgroup_path);
-        strcat(mem_usage_filename, MEM_USAGE_FILENAME);
-        result = ReadMemoryValueFromFile(mem_usage_filename, &temp);
-        if (result)
+        FILE *mountinfofile = fopen(PROC_MOUNTINFO_FILENAME, "r");
+        if (mountinfofile == nullptr)
+            goto done;
+
+        while (getline(&line, &lineLen, mountinfofile) != -1)
         {
-            if (temp > std::numeric_limits<size_t>::max())
+            // See man page of proc to get format for /proc/self/mountinfo file
+            char *separatorChar = strstr(line, " - ");
+            if (separatorChar == nullptr)
             {
-                *val = std::numeric_limits<size_t>::max();
+                assert(!"Failed to parse mount info file contents.");
+                goto done;
             }
-            else
+            char *filesystemTypeStart = separatorChar + strlen(" - ");
+            if (filesystemTypeStart == nullptr)
```

Comment:
This is never going to be true.  What did you want to do here?

Reply:
Doh. I was trying to be defensive against allocation failures, but there's no allocation here. I have removed this check.